File: internal/httpapi/scripture.go
package httpapi

import (
	"log"
	"net/http"
	"regexp"
	"strconv"
	"strings"

	"github.com/wiradigitalid/worship-presenter-web/internal/scripture"
)

func (s *Server) getScripture(w http.ResponseWriter, r *http.Request) {
	ref := strings.TrimSpace(r.URL.Query().Get("ref"))
	q := strings.TrimSpace(r.URL.Query().Get("q"))
	if ref == "" && q == "" {
		writeError(w, http.StatusBadRequest, "Missing ref or q query parameter")
		return
	}
	code, ok := s.resolveTranslation(w, r)
	if !ok {
		return
	}
	if ref != "" {
		passage, found := s.lookupScripture(ref, code)
		if !found {
			writeError(w, http.StatusNotFound, "Scripture reference not found")
			return
		}
		writeJSON(w, http.StatusOK, passage)
		return
	}
	names := s.loadBookNames(code)
	hits := scripture.SuggestBooks(q, names, scripture.AliasesFor(code), 20)
	suggestions := make([]map[string]string, 0, len(hits))
	for _, h := range hits {
		suggestions = append(suggestions, map[string]string{
			"name":       h.Name,
			"short_name": h.ShortName,
		})
	}
	writeJSON(w, http.StatusOK, map[string]any{
		"suggestions": suggestions,
		"translation": code,
	})
}

func (s *Server) resolveTranslation(w http.ResponseWriter, r *http.Request) (string, bool) {
	translationParam := strings.TrimSpace(r.URL.Query().Get("translation"))
	installed, err := s.listTranslationCodes()
	if err != nil {
		log.Printf("Error looking up scripture: %v", err)
		writeError(w, http.StatusInternalServerError, "Internal Server Error")
		return "", false
	}
	code := "KJV"
	if translationParam != "" {
		normalized := strings.ToUpper(translationParam)
		known := false
		for _, c := range installed {
			if c == normalized {
				known = true
				break
			}
		}
		if !known {
			writeError(w, http.StatusBadRequest, `Unknown bible translation "`+normalized+`"`)
			return "", false
		}
		code = normalized
	}
	var n int
	if err := s.DB.QueryRow(`SELECT COUNT(*) FROM bible_verses WHERE translation_code = ?`, code).Scan(&n); err != nil {
		writeError(w, http.StatusInternalServerError, "Internal Server Error")
		return "", false
	}
	if n == 0 {
		writeError(w, http.StatusServiceUnavailable,
			code+` corpus is empty. It ships at data/*/bible-translation/`+strings.ToLower(code)+`.json and is reconciled from that file on boot; check it with npm run corpus:verify.`)
		return "", false
	}
	return code, true
}

func (s *Server) listTranslationCodes() ([]string, error) {
	rows, err := s.DB.Query(`SELECT code FROM bible_translations ORDER BY code`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []string
	for rows.Next() {
		var c string
		if err := rows.Scan(&c); err != nil {
			return nil, err
		}
		out = append(out, c)
	}
	return out, rows.Err()
}

func (s *Server) lookupScripture(ref, code string) (map[string]string, bool) {
	bookPart, chapter, start, end, ok := scripture.ParseRef(ref)
	if !ok {
		return nil, false
	}
	bookID, canonical, ok := s.resolveBook(bookPart, code)
	if !ok {
		return nil, false
	}
	rows, err := s.DB.Query(
		`SELECT verse, verse_text FROM bible_verses
		  WHERE book_id = ? AND chapter = ? AND verse >= ? AND verse <= ? AND translation_code = ?
		  ORDER BY verse ASC`,
		bookID, chapter, start, end, code,
	)
	if err != nil {
		return nil, false
	}
	defer rows.Close()
	var texts []string
	for rows.Next() {
		var verse int
		var text string
		if err := rows.Scan(&verse, &text); err != nil {
			return nil, false
		}
		texts = append(texts, stripVerseMarkup(text))
	}
	if len(texts) == 0 {
		return nil, false
	}
	reference := canonical + " " + strconv.Itoa(chapter) + ":" + strconv.Itoa(start)
	if start != end {
		reference += "-" + strconv.Itoa(end)
	}
	return map[string]string{
		"reference":   reference,
		"text":        strings.Join(texts, " "),
		"translation": code,
	}, true
}

func (s *Server) resolveBook(bookPart, translation string) (int, string, bool) {
	names := s.loadBookNames(translation)
	id, canonical, ok := scripture.MatchBook(bookPart, names, scripture.AliasesFor(translation))
	return id, canonical, ok
}

func (s *Server) loadBookNames(translation string) []scripture.BookName {
	rows, err := s.DB.Query(
		`SELECT book_id, name, short_name FROM bible_book_names WHERE translation_code = ?`,
		translation,
	)
	if err == nil {
		defer rows.Close()
		var out []scripture.BookName
		for rows.Next() {
			var n scripture.BookName
			if err := rows.Scan(&n.ID, &n.Name, &n.ShortName); err != nil {
				break
			}
			out = append(out, n)
		}
		if len(out) > 0 {
			return out
		}
	}
	fallback, err := s.DB.Query(`SELECT id, name, short_name FROM bible_books`)
	if err != nil {
		return nil
	}
	defer fallback.Close()
	var out []scripture.BookName
	for fallback.Next() {
		var n scripture.BookName
		if err := fallback.Scan(&n.ID, &n.Name, &n.ShortName); err != nil {
			return out
		}
		out = append(out, n)
	}
	return out
}

func stripVerseMarkup(text string) string {
	text = regexp.MustCompile(`@\d+`).ReplaceAllString(text, "")
	return strings.Join(strings.Fields(text), " ")
}

File: src/lib/scripture.ts
import { getDb } from './db';

export type ScripturePassage = {
  reference: string;
  text: string;
  translation: string;
};

/**
 * Strip source markup like `@9was@7` / `@6…@5` — the export encoded the words
 * the 1611 translators supplied (printed italic in the KJV) this way. The
 * committed corpus at `data/en/bible-translation/kjv.json` is already clean and
 * `npm run corpus:verify` fails if a marker reappears, so this now guards
 * verses that reached the table by some other route.
 */
export function stripVerseMarkup(text: string): string {
  return text.replace(/@\d+/g, '').replace(/\s{2,}/g, ' ').trim();
}

type ParsedRef = {
  book: string;
  chapter: number;
  verseStart: number;
  verseEnd: number;
};

/** Strip placeholder prefixes operators often leave in the field (e.g. from UI copy). */
function normalizeScriptureInput(raw: string): string {
  return decodeURIComponent(raw)
    .replace(/\+/g, ' ')
    .trim()
    .replace(/^(?:e\.g\.|eg\.|example:)\s*/i, '')
    .trim();
}

/** Parse refs like `John 4:23`, `Song of Solomon 1:1`, `Hakim-hakim 2:16`. */
export function parseScriptureRef(raw: string): ParsedRef | null {
  const value = normalizeScriptureInput(raw);
  if (!value) return null;

  const colon = value.lastIndexOf(':');
  if (colon <= 0) return null;
  const before = value.slice(0, colon).trim();
  const after = value.slice(colon + 1).trim();
  const space = before.search(/\s+\S+$/);
  if (space < 0) return null;
  const book = before.slice(0, space).replace(/\s+/g, ' ').trim();
  const chapter = Number(before.slice(space + 1).trim());
  const span = after.match(/^(\d+)(?:\s*[-–]\s*(\d+)|\s*,\s*(\d+))?$/);
  if (!span) return null;
  const verseStart = Number(span[1]);
  const verseEnd = Number(span[2] || span[3] || span[1]);
  if (
    !book ||
    !Number.isInteger(chapter) ||
    !Number.isInteger(verseStart) ||
    !Number.isInteger(verseEnd) ||
    chapter <= 0 ||
    verseStart <= 0 ||
    verseEnd < verseStart
  ) {
    return null;
  }

  return { book, chapter, verseStart, verseEnd };
}

type BookName = { id: number; name: string; shortName: string };

function aliasesFor(translation: string): Array<{ alias: string; bookId: number }> {
  if (translation.trim().toUpperCase() !== 'KJV') return [];
  return [
    { alias: 'ps', bookId: 19 },
    { alias: 'psalm', bookId: 19 },
    { alias: 'psalms', bookId: 19 },
    { alias: 'sos', bookId: 22 },
    { alias: 'song of songs', bookId: 22 },
    { alias: 'song of solomon', bookId: 22 },
  ];
}

function normalizeBookKey(value: string): string {
  return value.trim().toLowerCase().replace(/\s+/g, ' ');
}

export type ScriptureBookSuggestion = {
  name: string;
  shortName: string;
};

const DEFAULT_SUGGEST_LIMIT = 20;

/** Prefix suggestions for the operator autocomplete. A complete ref is empty. */
export function suggestBooks(
  q: string,
  names: BookName[],
  aliases: Array<{ alias: string; bookId: number }>,
  limit = DEFAULT_SUGGEST_LIMIT
): ScriptureBookSuggestion[] {
  const cap =
    limit <= 0 || limit > DEFAULT_SUGGEST_LIMIT ? DEFAULT_SUGGEST_LIMIT : limit;
  if (parseScriptureRef(q)) return [];
  let input = normalizeBookKey(q);
  if (!input) return [];
  const fields = input.split(' ');
  if (fields.length >= 2 && /^\d+$/.test(fields[fields.length - 1] ?? '')) {
    const n = Number(fields[fields.length - 1]);
    if (Number.isInteger(n) && n > 0) {
      input = fields.slice(0, -1).join(' ');
    }
  }
  if (!input) return [];
  const byId = new Map(names.map((n) => [n.id, n]));
  const seen = new Set<number>();
  const out: ScriptureBookSuggestion[] = [];
  const consider = (key: string, book: BookName | undefined) => {
    if (!book) return;
    const k = normalizeBookKey(key);
    if (!k || (k !== input && !k.startsWith(input))) return;
    if (seen.has(book.id)) return;
    seen.add(book.id);
    out.push({ name: book.name, shortName: book.shortName });
  };
  for (const n of names) {
    consider(n.name, n);
    consider(n.shortName, n);
  }
  for (const a of aliases) {
    consider(a.alias, byId.get(a.bookId));
  }
  return out.slice(0, cap);
}

function matchBook(
  bookPart: string,
  names: BookName[],
  aliases: Array<{ alias: string; bookId: number }>
): { id: number; canonicalName: string } | null {
  const input = normalizeBookKey(bookPart);
  if (!input) return null;
  const byId = new Map(names.map((n) => [n.id, n.name]));
  const found: Array<{ id: number; canonical: string; keyLen: number }> = [];
  const consider = (key: string, bookId: number, canonical: string) => {
    const k = normalizeBookKey(key);
    if (!k) return;
    if (input === k || input.startsWith(`${k} `)) {
      found.push({ id: bookId, canonical, keyLen: k.length });
    }
  };
  for (const n of names) {
    consider(n.name, n.id, n.name);
    consider(n.shortName, n.id, n.name);
  }
  for (const a of aliases) {
    consider(a.alias, a.bookId, byId.get(a.bookId) ?? '');
  }
  if (found.length === 0) return null;
  const max = Math.max(...found.map((c) => c.keyLen));
  const longest = found.filter((c) => c.keyLen === max);
  const ids = new Set(longest.map((c) => c.id));
  if (ids.size !== 1) return null;
  const winner = longest[0];
  return {
    id: winner.id,
    canonicalName: winner.canonical || byId.get(winner.id) || bookPart,
  };
}

function loadBookNames(translationCode: string): BookName[] {
  const db = getDb();
  try {
    const named = db
      .prepare(
        `SELECT book_id AS id, name, short_name AS shortName
           FROM bible_book_names WHERE translation_code = ?`
      )
      .all(translationCode) as BookName[];
    if (named.length > 0) return named;
  } catch {
    // Tests and databases created before bible_book_names still read bible_books.
  }
  return db
    .prepare(`SELECT id, name, short_name AS shortName FROM bible_books`)
    .all() as BookName[];
}

function resolveBookId(
  bookName: string,
  translationCode: string
): { id: number; canonicalName: string } | null {
  return matchBook(
    bookName,
    loadBookNames(translationCode),
    aliasesFor(translationCode)
  );
}

/** True when the named translation holds no verses in the table. */
export function isBibleTranslationEmpty(translationCode: string): boolean {
  const code = translationCode.trim().toUpperCase();
  const db = getDb();
  const row = db
    .prepare(
      `SELECT COUNT(*) AS n FROM bible_verses WHERE translation_code = ?`
    )
    .get(code) as { n: number };
  return !row || row.n === 0;
}

/**
 * Look up scripture text for a reference in the named translation. For deck
 * theme/verse slides, do NOT call this — use rundown-supplied scripture only.
 */
export function lookupScripture(
  ref: string,
  translationCode: string
): ScripturePassage | null {
  const code = translationCode.trim().toUpperCase();
  const parsed = parseScriptureRef(ref);
  if (!parsed) return null;

  const matched = resolveBookId(parsed.book, code);
  if (matched == null) return null;

  const db = getDb();
  const rows = db
    .prepare(
      `SELECT verse, verse_text FROM bible_verses
       WHERE book_id = ? AND chapter = ? AND verse >= ? AND verse <= ?
         AND translation_code = ?
       ORDER BY verse ASC`
    )
    .all(
      matched.id,
      parsed.chapter,
      parsed.verseStart,
      parsed.verseEnd,
      code
    ) as {
    verse: number;
    verse_text: string;
  }[];

  if (rows.length === 0) return null;

  const text = rows.map((r) => stripVerseMarkup(r.verse_text)).join(' ');
  const reference =
    parsed.verseStart === parsed.verseEnd
      ? `${matched.canonicalName} ${parsed.chapter}:${parsed.verseStart}`
      : `${matched.canonicalName} ${parsed.chapter}:${parsed.verseStart}-${parsed.verseEnd}`;

  return { reference, text, translation: code };
}

/** Book suggestions for the named translation. Operator autocomplete only. */
export function suggestScriptureBooks(
  q: string,
  translationCode: string
): ScriptureBookSuggestion[] {
  const code = translationCode.trim().toUpperCase();
  return suggestBooks(q, loadBookNames(code), aliasesFor(code));
}
